### PR TITLE
Implementations for atomic load, store and exchange operations

### PR DESCRIPTION
**Pull request by adarshyoga**
- [x] Have you provided a meaningful PR description?

This PR adds overloads and intrinsics for load, store and exchange atomic operations. It also contains unit tests that verify the functionality and the failure cases for these operations.

- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: numba_dpex/tests/experimental/kernel_iface/spv_overloads/test_atomic_cmp_exchg.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
+import dpnp
```
Comment (diptorupd) [inline numba_dpex/tests/experimental/kernel_iface/spv_overloads/test_atomic_cmp_exchg.py:1]: License header

File: numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_atomic_ref_overloads.py [hunk 2]
```diff
@@ -209,6 +209,157 @@ def codegen(context, builder, sig, args):
...
+
+
+@intrinsic(target=DPEX_KERNEL_EXP_TARGET_NAME)
+def _intrinsic_store(
+    ty_context, ty_atomic_ref, ty_val
+):  # pylint: disable=unused-argument
+    sig = types.void(ty_atomic_ref, ty_val)
+
+    def _intrinsic_store_gen(context, builder, sig, args):
+        _store_exchange_intrisic_helper(
```
Comment (diptorupd) [inline numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_atomic_ref_overloads.py:324]: Why name `store_exchange_intrinsic_helper` v/s `store_intrinsic_helper`?

Comment (adarshyoga) [reply numba_dpex/experimental/_kernel_dpcpp_spirv_overloads/_atomic_ref_overloads.py:324]: It is used by both store and exchange functions

**Comment by diptorupd** [comment]
@adarshyoga the crash is reproducible if you use `ONEAPI_DEVICE_SELECTOR=opencl:cpu`

I ran it with gdb and here is the backtrace:

```bash
0x00007ffff5f325f5 in std::_Function_handler<void (sycl::_V1::handler&), DPCTLQueue_SubmitRange::{lambda(sycl::_V1::handler&)#1}>::_M_invoke(std::_Any_data const&, sycl::_V1::handler&) ()
   from /home/diptorupd/miniconda3/envs/dpex-devel/lib/python3.11/site-packages/dpctl/libDPCTLSyclInterface.so.0
(gdb) bt
#0  0x00007ffff5f325f5 in std::_Function_handler<void (sycl::_V1::handler&), DPCTLQueue_SubmitRange::{lambda(sycl::_V1::handler&)#1}>::_M_invoke(std::_Any_data const&, sycl::_V1::handler&) ()
   from /home/diptorupd/miniconda3/envs/dpex-devel/lib/python3.11/site-packages/dpctl/libDPCTLSyclInterface.so.0
#1  0x00007ffff5b0ce36 in sycl::_V1::detail::queue_impl::submit_impl(std::function<void (sycl::_V1::handler&)> const&, std::shared_ptr<sycl::_V1::detail::queue_impl> const&, std::shared_ptr<sycl::_V1::detail::queue_impl> const&, std::shared_ptr<sycl::_V1::detail::queue_impl> const&, sycl::_V1::detail::code_location const&, std::function<void (bool, bool, sycl::_V1::event&)> const*)
    () from /home/diptorupd/miniconda3/envs/dpex-devel/lib/python3.11/site-packages/dpctl/../../../libsycl.so.6
#2  0x00007ffff5b0c406 in sycl::_V1::detail::queue_impl::submit(std::function<void (sycl::_V1::handler&)> const&, std::shared_ptr<sycl::_V1::detail::queue_impl> const&, sycl::_V1::detail::code_location const&, std::function<void (bool, bool, sycl::_V1::event&)> const*) () from /home/diptorupd/miniconda3/envs/dpex-devel/lib/python3.11/site-packages/dpctl/../../../libsycl.so.6
#3  0x00007ffff5b0c3c5 in sycl::_V1::queue::submit_impl(std::function<void (sycl::_V1::handler&)>, sycl::_V1::detail::code_location const&) ()
   from /home/diptorupd/miniconda3/envs/dpex-devel/lib/python3.11/site-packages/dpctl/../../../libsycl.so.6
#4  0x00007ffff5f2e6bc in DPCTLQueue_SubmitRange () from /home/diptorupd/miniconda3/envs/dpex-devel/lib/python3.11/site-packages/dpctl/libDPCTLSyclInterface.so.0
#5  0x00007fffa75861fd in numba_dpex::experimental::launcher::call_kernel[abi:v16][abi:c8tJTC_2fWQAlzW1yBDkop6GEOEUMEOYSPGuIQMViAQ3iQ8IbKQIMbwoOGNoQDDWwQR1NHAS3lQ9XgSucwm4pgLNTQs00pSK3QBAA_3d](type_28KernelDispatcher_28_3cfunction_20test_load_store_fn::_3clocals_3e::_kernel_20at_200x7fff4613ba60_3e_29_29, Range_3c1_3e, UniTuple<DpnpNdArray<double, 1, C, opencl_cpu>, 2>) ()
#6  0x00007fffa75866ca in cpython::numba_dpex::experimental::launcher::call_kernel[abi:v16][abi:c8tJTC_2fWQAlzW1yBDkop6GEOEUMEOYSPGuIQMViAQ3iQ8IbKQIMbwoOGNoQDDWwQR1NHAS3lQ9XgSucwm4pgLNTQs00pSK3QBAA_3d](type_28KernelDispatcher_28_3cfunction_20test_load_store_fn::_3clocals_3e::_kernel_20at_200x7fff4613ba60_3e_29_29, Range_3c1_3e, UniTuple<DpnpNdArray<double, 1, C, opencl_cpu>, 2>) ()
#7  0x00007ffff357da25 in compile_and_invoke(Dispatcher*, _object*, _object*, _object*) ()
```

**Comment by adarshyoga** [comment]
> @adarshyoga the crash is reproducible if you use `ONEAPI_DEVICE_SELECTOR=opencl:cpu`
> 
> I ran it with gdb and here is the backtrace:
> 
> ```shell
> 0x00007ffff5f325f5 in std::_Function_handler<void (sycl::_V1::handler&), DPCTLQueue_SubmitRange::{lambda(sycl::_V1::handler&)#1}>::_M_invoke(std::_Any_data const&, sycl::_V1::handler&) ()
>    from /home/diptorupd/miniconda3/envs/dpex-devel/lib/python3.11/site-packages/dpctl/libDPCTLSyclInterface.so.0
> (gdb) bt
> #0  0x00007ffff5f325f5 in std::_Function_handler<void (sycl::_V1::handler&), DPCTLQueue_SubmitRange::{lambda(sycl::_V1::handler&)#1}>::_M_invoke(std::_Any_data const&, sycl::_V1::handler&) ()
>    from /home/diptorupd/miniconda3/envs/dpex-devel/lib/python3.11/site-packages/dpctl/libDPCTLSyclInterface.so.0
> #1  0x00007ffff5b0ce36 in sycl::_V1::detail::queue_impl::submit_impl(std::function<void (sycl::_V1::handler&)> const&, std::shared_ptr<sycl::_V1::detail::queue_impl> const&, std::shared_ptr<sycl::_V1::detail::queue_impl> const&, std::shared_ptr<sycl::_V1::detail::queue_impl> const&, sycl::_V1::detail::code_location const&, std::function<void (bool, bool, sycl::_V1::event&)> const*)
>     () from /home/diptorupd/miniconda3/envs/dpex-devel/lib/python3.11/site-packages/dpctl/../../../libsycl.so.6
> #2  0x00007ffff5b0c406 in sycl::_V1::detail::queue_impl::submit(std::function<void (sycl::_V1::handler&)> const&, std::shared_ptr<sycl::_V1::detail::queue_impl> const&, sycl::_V1::detail::code_location const&, std::function<void (bool, bool, sycl::_V1::event&)> const*) () from /home/diptorupd/miniconda3/envs/dpex-devel/lib/python3.11/site-packages/dpctl/../../../libsycl.so.6
> #3  0x00007ffff5b0c3c5 in sycl::_V1::queue::submit_impl(std::function<void (sycl::_V1::handler&)>, sycl::_V1::detail::code_location const&) ()
>    from /home/diptorupd/miniconda3/envs/dpex-devel/lib/python3.11/site-packages/dpctl/../../../libsycl.so.6
> #4  0x00007ffff5f2e6bc in DPCTLQueue_SubmitRange () from /home/diptorupd/miniconda3/envs/dpex-devel/lib/python3.11/site-packages/dpctl/libDPCTLSyclInterface.so.0
> #5  0x00007fffa75861fd in numba_dpex::experimental::launcher::call_kernel[abi:v16][abi:c8tJTC_2fWQAlzW1yBDkop6GEOEUMEOYSPGuIQMViAQ3iQ8IbKQIMbwoOGNoQDDWwQR1NHAS3lQ9XgSucwm4pgLNTQs00pSK3QBAA_3d](type_28KernelDispatcher_28_3cfunction_20test_load_store_fn::_3clocals_3e::_kernel_20at_200x7fff4613ba60_3e_29_29, Range_3c1_3e, UniTuple<DpnpNdArray<double, 1, C, opencl_cpu>, 2>) ()
> #6  0x00007fffa75866ca in cpython::numba_dpex::experimental::launcher::call_kernel[abi:v16][abi:c8tJTC_2fWQAlzW1yBDkop6GEOEUMEOYSPGuIQMViAQ3iQ8IbKQIMbwoOGNoQDDWwQR1NHAS3lQ9XgSucwm4pgLNTQs00pSK3QBAA_3d](type_28KernelDispatcher_28_3cfunction_20test_load_store_fn::_3clocals_3e::_kernel_20at_200x7fff4613ba60_3e_29_29, Range_3c1_3e, UniTuple<DpnpNdArray<double, 1, C, opencl_cpu>, 2>) ()
> #7  0x00007ffff357da25 in compile_and_invoke(Dispatcher*, _object*, _object*, _object*) ()
> ```

Thats great! Thanks for fixing this and adding the license.
So, was the crash occurring when using non-float32 dtypes of the input arrays with `opencl:cpu`?

Also, is there anything more needed for this PR? Do the two commits need to be squashed?

File: numba_dpex/tests/experimental/kernel_iface/spv_overloads/test_atomic_load_store.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+
+@pytest.mark.parametrize(
+    "dtype_param",
+    list_of_supported_dtypes
+    + [
+        dpnp.float32,
+        pytest.param(
+            dpnp.float64,
+            marks=pytest.mark.skip(
```
Comment (diptorupd) [inline numba_dpex/tests/experimental/kernel_iface/spv_overloads/test_atomic_load_store.py:31]: Is this not skipping on all devices and not just on opencl:cpu?

**Review by diptorupd — APPROVED** [review]

**Pull request merged**